stream flash API

hakonfam opened the pull request:
Extract stream write to flash functionality from flash_img to a general module.

Why:
- This code is useful in non-mcuboot and non-dfu contexts as well

Goals:
- Don't break API of flash_img

hakonfam committed:
storage: add stream flash library

This library supports stream writes to flash with
optinal progressive erase.

This module is a direct copy of the functionality found in
subsys/dfu/img_util/flash_img.c

Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>

hakonfam committed:
dfu: use stream_flash in flash_img

Leverage strem_flash in flash_img.c to reduce code size.

Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>

mbolivar commented on include/storage/fbw.h:
```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2017 Nordic Semiconductor ASA
```
```suggestion
 * Copyright (c) 2017, 2020 Nordic Semiconductor ASA
```

mbolivar replied:
Likewise elsewhere, I think

hakonfam replied:
OK, even though this is a copy-paste with a simple rename of old code? I'm not well known in the world of copyright :)

mbolivar replied:
Me neither, but it seems like you're adding various pieces of documentation etc. that are new. Not sure if there's a well defined policy.

mbolivar commented on subsys/storage/flash_buffered_write/fbw.c:
```diff
@@ -0,0 +1,210 @@
+	ctx->off_last = -1;
+#endif
+
+	return flash_area_open(area_id,
```
Can you explain the decision to move the flash_area_open() / flash_area_close() calls into fbw.c?

Naively, I would have guessed that this would be the responsibility of any callers to fbw_write() or other I/O routines. That is, the fbw.h API would just assume any flash it's asked to do I/O on would be ready for it.

I'm also a bit surprised not to see a `struct device*` in the flash context structure and the flash API being used to do the actual erasing/writing. Is there a reason to keep using flash areas here?

I ask because I've honestly always had my doubts about flash_map.h. It was basically copy/pasted into Zephyr from Mynewt code, and I didn't think it was a good idea at the time, either, since it doesn't play natively with the Zephyr driver model.

hakonfam replied:
TBH I didn't consider the rationale behind the already made choices for this module, I simply copy-pasted the functionality found in flash_img.c to allow for more re-use.

The user of this module (of which we have identified 3 until now) all have the same use case, pass a pointer to some flash area and a length, and write it to a different part of flash, internal or external.

The interface provided by this module is sufficient to solve those use cases, while requiring a very small amount of boilerplate code (opening closing flash areas etc.)

nvlsianpu replied:
Maybe good idea would be to transfer flash_area by the user - this will make this API the most flexible. Flash area might be then defined either in the build-time or be even created in the run-time.

hakonfam replied:
I agree

mbolivar replied:
I don't think we should stick to the current flash area API when we generalize this, to be honest.

The current `<storage/flash_map.h>` API has unnecessary inefficiencies in all of its I/O operations. The read, write, etc. routines all call device_get_binding() for every I/O op, which in the worst case is iterating through all of the devices in the system.

It also hard-codes its device IDs (`SOC_FLASH_0_ID` and `SPI_FLASH_0_ID`) instead of just using the Zephyr driver model the way it was intended with struct devices.

I would be against adding a new top-level API that relies on flash map for this reason.

hakonfam replied:
I'm working on a variant now which has the `struct device *fdev` member in the `fbw_ctx` instead, this will not use the flash_map at all, but the API of `flash_img.c` will not change, so the `flash_map` handling will be moved there as you proposed somewhere else.

nvlsianpu replied:
@mbolivar - I think it is easy to improve flash_area API implementation regards this, for instance inserting the driver reference into flash_area struct... (struct need to be runtime then...).

mbolivar commented on include/storage/fbw.h:
```diff
@@ -0,0 +1,102 @@
+extern "C" {
+#endif
+
+struct fbw_ctx {
```
The API only uses pointers to fbw_ctx, so I'm assuming the structure contents are here because it's caller-allocated.

If so, that should be documented when you're ready to take this out of draft status. Something like this maybe:

```suggestion
/**
 * @brief Context structure for buffered flash writes.
 * Users should treat these structures as opaque values and only interact
 * with them through the below API.
 */
struct fbw_ctx {
```

hakonfam replied:
> The API only uses pointers to fbw_ctx, so I'm assuming the structure contents are here because it's caller-allocated.

I'm not sure what you mean by 'structure contents'. But yes, I agree that this should be documented as you propose.

mbolivar replied:
> I'm not sure what you mean by 'structure contents'.

I mean showing the members inside of struct fbw_ctx instead of defining it as an opaque type.

mbolivar commented on include/storage/fbw.h:
```diff
@@ -0,0 +1,102 @@
+};
+
+/**
+ * @brief Initialize context needed for buffered write to flash.
```
When you are ready to take this out of draft status, I think it'd be good to mark up these API functions according to the new [terminology](https://docs.zephyrproject.org/latest/reference/index.html#terminology) introduced by @pabigot 

For example, I guess this one is pre-kernel-ok and not reschedule.

hakonfam replied:
I will make sure to mark them up (Y)

hakonfam replied:
I'm not able to find any example usage of those tags, could you point me to a file where this is already done?

mbolivar replied:
You're a guinea pig :)

nvlsianpu replied:
someone have to be the firs one

hakonfam replied:
I tried adding some tags. Really unsure about the syntax that should be used (is there any?).

Also, I'm not sure if flash operations are available pre-kernel? If not then I guess all my proposed tags are wrong, so please check those xD

mbolivar replied:
> When you are ready to take this out of draft status, I think it'd be good to mark up these API functions according to the new [terminology](https://docs.zephyrproject.org/latest/reference/index.html#terminology) introduced by @pabigot

Also CC @carlescufi -- can we add this to the API agenda for next week as a test run?

carlescufi replied:
API meeting:

Suggestion from @nashif:
- Use a doxygen alias to map something like `api-tags` to a regular Doxygen `note` that can be parsed by breathe.

@mbolivar-nordic  @hakonfam: let's  use `note` for now, and we will revisit this anyway once the above is resolved

hakonfam replied:
Added @note, please check if it looks correct

mbolivar-nordic replied:
I would vote for `@note pre-kernel-ok` or so, but will leave this up to @carlescufi and @pabigot

pabigot replied:
I have no strong opinions about the style associated with documenting API behavior, except that I would avoid making things too strained by attempting to prepare for some future machine processing that is unlikely to exist anytime soon.  "This function is isr-ok and not sleep." covers the core need, with or without a preceding `@note`.  (Said future tooling should convert `sleep` to [sleep](https://docs.zephyrproject.org/latest/reference/index.html#sleep) to make clear it's being used as an adjective/attribute, not a verb.)

For the content: [pre-kernel-ok](https://docs.zephyrproject.org/latest/reference/index.html#pre-kernel-ok) is one of the less useful tags.  It's really only important to know if the API is something that could reasonably be expected to be invoked before startup switches from the interrupt stack to a thread stack.  In that case it almost certainly would be paired with [isr-ok](https://docs.zephyrproject.org/latest/reference/index.html#isr-ok).

Developers also need to be aware that both `pre-kernel-ok` and `isr-ok` in isolation can be satisfied by having the function return an error if called from a context where they aren't allowed to sleep.   Both should be paired with `not sleep` to suggest that the function actually does its job when invoked from a non-thread context.

AFAICT all flash drivers initialize post-kernel, so if this function requires a flash device it's pretty unlikely that `pre-kernel-ok` is relevant.

Also don't forget to document expectations for the `cb` parameter: does this API expect it to be invoked from an interrupt handler?  That might be best documented at the typedef for `fbw_callback_t`, which currently says nothing about what it is or how it's used.

hakonfam replied:
I removed the `pre-kernel-ok` for all funtions except `fbw_bytes_written` which has both `pre-kernel-ok` and `isr-ok`. 

As for `fbw_callback_t` I added docs to that typedef now, but it is not clear to me how I should explain where the "API expect it to be invoked from an interrupt handler", so if we want that you would have to help me with the phrasing.

pabigot replied:
From what I can tell `fbw_oper` inherits the API behavior of `flash_oper` given whatever flash device is provided in the context.  Some will be **sleep**, few if any will be **isr-ok**.

hakonfam replied:
OK, then I suggest that we remove the API-tags altogether, as it is difficult to give any information of value.

mbolivar commented on include/storage/fbw.h:
```diff
@@ -0,0 +1,102 @@
+	const struct flash_area *flash_area;
+	size_t bytes_written;
+	u16_t buf_bytes;
+#ifdef CONFIG_IMG_ERASE_PROGRESSIVELY
```
This is a flash_img.h config option; right? I think it shouldn't be in this API if so. Would it need to get renamed as part of making this API generic?

hakonfam replied:
Yes, I made a separate Kconfig for this module, will change to this.

nvlsianpu replied:
CONFIG_FBW_ERASE_PROGRESSIVELY
Wuld be great if IMG_ERASE_PROGRESSIVELY will be kept as selector for the  new keyword - this will ensure API backward compatibility.

hakonfam replied:
I planned on keeping the backward compatibility by having IMG_ERASE_PROGRESSIVELY select FBW_ERASE_PROGRESSIVELY

mbolivar commented on include/storage/fbw.h:
```diff
@@ -0,0 +1,102 @@
+size_t fbw_bytes_written(struct fbw_ctx *ctx);
+
+/**
+ * @brief  Process input buffers to be written to the image slot 1. flash
```
These need a bit of a cleanup to remove flash_img specific language, right?

hakonfam replied:
Correct! Sorry for the sloppy work, I just threw this out so I could get your review before office time tomorrow :)

mbolivar replied:
Not sloppy at all! I really appreciate the handoff -- I know you're posting this as a draft.

mbolivar commented on include/storage/fbw.h:
```diff
@@ -0,0 +1,102 @@
+ * A final call to this function with flush set to true
+ * will write out the remaining block buffer to flash. Since flash is written to
+ * in blocks, the contents of flash from the last byte written up to the next
+ * multiple of CONFIG_IMG_BLOCK_BUF_SIZE is padded with 0xff.
```
Now that this API is being generalized, does the padding value need to be a parameter to the function or left up to the flash device instead of hard-coded to 0xff?

hakonfam replied:
That makes sense, especially as this can be used against any type of flash device.

hakonfam replied:
I'm not padding the buffer anymore, since the pages are erased anyway this does not makes sense IMO.

nvlsianpu reviewed (commented):
As I saw, Marti already did quite comprehensive review.

Name - I'm fine with `FBW` name, I to less descriptive might be `flash_bw` as well.

Regard FBW_ERASE_PROGRESSIVELY in this module:
Would be great if IMG_ERASE_PROGRESSIVELY will be kept as selector for the new keyword - this will ensure API backward compatibility.

hakonfam commented:
@mbolivar  @nvlsianpu  The current version should be ready for a new round of reviews.

The changes are:
- Use `struct device *` member instead of `flash_area`.
- Allow `offset` and `size` to be given as input (to allow subsets of flash devices to be used)
- Write new unit tests, keep flash_img tests

mbolivar commented on include/storage/fbw.h:
```diff
@@ -0,0 +1,113 @@
+ * @param fdev Flash device to operate on
+ * @param buf Write buffer
+ * @param buf_len Length of write buffer
+ * @param offset Offset within flash device
```
Would it be correct to write this as:

```suggestion
 * @param offset Offset within flash device to start writing to
```

mbolivar replied:
Also, what kind of erase state does the sector containing this offset have to be in?

hakonfam replied:
The same as the other ones. If `FWB_ERASE=y` then they will be erased before written to. If `FWB_ERASE=n` then they have to be erased.

mbolivar commented on include/storage/fbw.h:
```diff
@@ -0,0 +1,113 @@
+ * @param buf Write buffer
+ * @param buf_len Length of write buffer
+ * @param offset Offset within flash device
+ * @param size Size of region within flash device
```
Looking at this with fresh eyes, does this mean the maximum number of bytes to write is `size - offset`? Or is it `size`?

Might want to clarify in the API docs here so lazy folks like me don't have to read the code ;)

hakonfam replied:
In my original version I was providing the `end` address instead of the size. Whilst making it more clear what the user should expect, I think it is more convenient to use the size, as this information is typically more available (start - size of partitions is more common info than start - end address).

Also I'm not sure if the `send 0 to get the total size` functionality is really that useful, so feedback on that is welcome.

mbolivar replied:
> Also I'm not sure if the `send 0 to get the total size` functionality is really that useful, so feedback on that is welcome.

+1 to this and the change to make it the size of the buffer area

mbolivar commented on include/storage/fbw.h:
```diff
@@ -0,0 +1,113 @@
+ * Erase the flash area progressively
+ *
+ * This function erases a flash page to which an offset belongs if this page
+ * wasn't erased before by this module.
```
This part I find a bit confusing when trying to imagine what this API does from only these doxygen comments. Some questions that come into my mind when looking at this API with a blank slate:

- Does that mean I have to call fbw_init() again to allow an erase of the same sector? Will this return an error if I try to erase something again, or is behavior undefined?
- Does this API somehow remember which sectors it has erased? If so, where does the memory for that get allocated from? If not, does the context just remember a most recent sector written to or something?

hakonfam replied:
I tried to rephrase the docs. Note that the only reason this is exposed is because this is the exact functionality required by mcuboot flash image writers, and I prefer providing a "weird" function rather than duplicating code.

mbolivar commented on subsys/storage/flash_buffered_write/fbw.c:
```diff
@@ -0,0 +1,181 @@
+ */
+
+#define LOG_MODULE_NAME fbw
+#define LOG_LEVEL CONFIG_fbw_LOG_LEVEL
```
Is that right?

hakonfam replied:
Changed to FBW

mbolivar commented on tests/subsys/storage/flash_buffered_write/src/main.c:
```diff
@@ -0,0 +1,302 @@
+	ztest_test_suite(lib_fbw_test,
+	     ztest_unit_test(test_fbw__init),
+	     ztest_unit_test(test_fbw__write),
+	     ztest_unit_test(test_fbw__write_cross_buf_border),
```
Just curious... what's up with the double underscores here? Seems weird.

hakonfam replied:
Just something I used to do back in the day to make it easier to read what tests were passing on an old CI system we had, will change it back to how its done in the other tests.

mbolivar commented on include/storage/fbw.h:
```diff
@@ -0,0 +1,113 @@
+ * @param buf_len Length of write buffer
+ * @param offset Offset within flash device
+ * @param size Size of region within flash device
+ * @param cb Callback to be invoked on completed write operations
```
I think some API guarantees about when exactly this gets called and with what arguments would be good to add.

hakonfam replied:
See new proposal.

mbolivar reviewed (commented):
I'm going to focus on the docstrings for now since this is a new API. I think the fbw_ctx structure looks sane.

mbolivar commented:
> Write new unit tests, keep flash_img tests

Thanks a lot for these! Very nice.

hakonfam commented:
Added DNM because of west.yml update with pull ref

https://github.com/zephyrproject-rtos/mcumgr/pull/15

jfischer-phytec-iot commented on subsys/storage/flash_buffered_write/fbw.c:
```diff
@@ -0,0 +1,181 @@
+	}
+
+	ctx->last_erased_page_start_offset = page.start_offset;
+	LOG_INF("Erasing page at offset 0x%08lx", (long)page.start_offset);
```
`LOG_DBG`?

jfischer-phytec-iot commented on include/storage/fbw.h:
```diff
@@ -0,0 +1,131 @@
+ *           the length of the written data, and the offset written to is sent
+ *           as an argument to the callback.
+ *
+ * @return  0 on success, negative errno code on fail
```
positive is not defined! Same below on few places.

hakonfam replied:
I'm not sure I understand what you mean, should I explicitly state that positive return values are not used?

jfischer-phytec-iot replied:
maybe "0 on success, non-zero on failure"?

hakonfam replied:
IMO it is better to state that the return value is negative. This way a caller which forwards the error code knows if it should be negated or not, to be made negative.

Also this seems to be the idiomatic way to describe this type of return code, it is used around 330 times AFAICS.

pabigot replied:
I'm pushing for "non-negative for success, negative for error" to be used generally.  This provides a binary decision to the caller (no question about success/failure) while allowing for distinguishing types of success, even if the specific values aren't part of the public API.  It's a nice way to get internal state information out of functions in unit tests.

hakonfam replied:
If we go for this we can no longer use `if (ret)` which is a common way of checking if the return value indicates success or not

```
➜  zephyr git:(fbw) ✗ rg -t c "if \((ret|rc|retval|ret_bytes)\)" | wc   
    655    2631   29820
➜  zephyr git:(fbw) ✗ rg -t c "if \((ret|rc|retval|ret_bytes)[^\)]" | wc 
   1903   11740  102000
```

hakonfam replied:
Updated to your proposal @pabigot

pabigot replied:
It's still common in existing code, but it's also in violation of MISRA (treating an `int` value as a boolean), which is why you may see new code use `if (ret != 0)`.  Then `if (ret < 0)` becomes an easy difference.

hakonfam replied:
Good point. I changed all tests from flash driver to `if (rc != 0)` as the API does not define whether positive value is success or not.

When testing the `rc` of `fbw_erase` I use `if (rc < 0)` as positive is documented as success in this case.

jfischer-phytec-iot commented on subsys/storage/flash_buffered_write/fbw.c:
```diff
@@ -0,0 +1,181 @@
+#define LOG_MODULE_NAME FBW
+#define LOG_LEVEL CONFIG_FBW_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
```
`LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_FBW_LOG_LEVEL);`

jfischer-phytec-iot commented on subsys/storage/flash_buffered_write/fbw.c:
```diff
@@ -0,0 +1,181 @@
+	size_t write_addr = ctx->offset + ctx->bytes_written;
+
+
+#ifdef CONFIG_FBW_ERASE
```
You can use  `if (IS_ENABLED(CONFIG_FBW_ERASE)) {`.

hakonfam replied:
AFAIK this does not work well with latest GCC versions, but I will double check.

hakonfam replied:
Checked with recent GCC version and it seems to work fine.

jfischer-phytec-iot commented on subsys/storage/flash_buffered_write/fbw.c:
```diff
@@ -0,0 +1,181 @@
+	ctx->offset = offset;
+	ctx->available = (size == 0 ? total_size - offset : size);
+	ctx->callback = cb;
+#ifdef CONFIG_FBW_ERASE
```
You can use ` if (IS_ENABLED(CONFIG_FBW_ERASE)) {`

hakonfam replied:
This actually fails with `gcc version 9.2.1 20200130 (Arch Linux 9.2.1+20200130-2)` so for now I only use `IS_ENABLED` earlier in this file. We might want to consider to only use `ifdef` to be more consistent?

hakonfam replied:
Updated. I removed the `#ifdef` surrounding the definition of the fbw_erase funtion as well.

hakonfam replied:
I'm not able to get `IS_ENABLED` to work the same way as `#ifdef`. See the error below.

```
fbw.c: In function 'fbw_init':
fbw.c:179:6: error: 'struct fbw_ctx' has no member named 'last_erased_page_start_offset'
   ctx->last_erased_page_start_offset = -1;
      ^~
```

From the lines 
```
	if (IS_ENABLED(CONFIG_FBW_ERASE)) {
		ctx->last_erased_page_start_offset = -1;
	}
```

Where as using `#ifdef CONFIG_FBW_ERASE` does not give an error.

hakonfam replied:
I guess this is because the definition of the struct is in a different file, and has the `#ifdef` surrounding the inclusion of this member variable.

jfischer-phytec-iot replied:
sad. Sorry I missed it.

jfischer-phytec-iot commented on tests/subsys/storage/flash_buffered_write/prj.conf:
```diff
@@ -0,0 +1,13 @@
+CONFIG_ZTEST=y
+CONFIG_FLASH=y
+CONFIG_FLASH_PAGE_LAYOUT=y
+CONFIG_MPU_ALLOW_FLASH_WRITE=y
```
Seems like it would be better to have an overlay with `CONFIG_MPU_ALLOW_FLASH_WRITE=y` for the platforms which depends on it.

hakonfam replied:
I agree, will change the logics

mbolivar commented:
Andrzej Puzdrowski <notifications@github.com> writes:

> nvlsianpu commented on this pull request.
>
>
>
>> +{
> +	return ctx->bytes_written;
> +}
> +
> +int fbw_init(struct fbw_ctx *ctx, u8_t *buf, size_t buf_len, u8_t area_id,
> +	     size_t offset)
> +{
> +	ctx->buf = buf;
> +	ctx->buf_len = buf_len;
> +	ctx->bytes_written = 0;
> +	ctx->buf_bytes = 0U;
> +#ifdef CONFIG_FBW_ERASE_PROGRESSIVELY
> +	ctx->off_last = -1;
> +#endif
> +
> +	return flash_area_open(area_id,
>
> @mbolivar - I think it is easy to improve flash_area API
> implementation regards this, for instance inserting the driver
> reference into flash_area struct... (struct need to be runtime
> then...).

I don't understand what "struct need to be runtime" means, sorry.

Can you change the flash_map API compatibly, without breaking existing users?

>
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/zephyrproject-rtos/zephyr/pull/23290#discussion_r389542603

nvlsianpu commented on include/storage/fbw.h:
```diff
@@ -0,0 +1,131 @@
+	struct device *fdev; /* Flash device */
+	size_t bytes_written; /* Number of bytes written to flash */
+	size_t offset; /* Offset from base of flash device to write area */
+	size_t available; /* Available bytes in write area */
```
area_size or partition_size, would be more self descriptive name

nvlsianpu reviewed (approved):
Code looks good to me - worth to switch this to regular PR.

nvlsianpu commented:
@mbolivar 
for my flash_area related thought see: https://gist.github.com/nvlsianpu/77e02037d0403136bdef528504d0f1da

aescolar commented on tests/subsys/dfu/img_util/prj_native_posix.conf:
```diff
@@ -1,5 +1,4 @@
 CONFIG_ZTEST=y
-CONFIG_STDOUT_CONSOLE=y
```
This seems quite unrelated, or?

hakonfam replied:
I agree, I just tried to be sneaky and fix an annoying warning, I will remove it.

aescolar replied:
Don't be sneaky ;), that will confuse people looking at the history. Just send a separate commit and/or separate PR

jfischer-phytec-iot commented on subsys/storage/flash_buffered_write/fbw.c:
```diff
@@ -0,0 +1,182 @@
+	rc = flash_write(ctx->fdev, write_addr, ctx->buf, ctx->buf_bytes);
+	flash_write_protection_set(ctx->fdev, true);
+
+	if (rc) {
```
not consistent, below it is `if (rc != 0) {`

hakonfam replied:
Changed to `if (rc)` everywhere.

mbolivar-nordic reviewed (commented):
LGTM modulo the open question about API tags.

@carlescufi does this need to be added to the API page? I guess it would have 'experimental' status.

carlescufi commented:
> LGTM modulo the open question about API tags.
> 
> @carlescufi does this need to be added to the API page? I guess it would have 'experimental' status.

Yes, this is a new API effectively, so it needs to be added. @hakonfam please add an entry in `doc/reference/stability.rst` and also include it in the doc reference, probably in `doc/reference/storage/`

carlescufi commented:
@de-nordic please review

hakonfam commented:
> > LGTM modulo the open question about API tags.
> > @carlescufi does this need to be added to the API page? I guess it would have 'experimental' status.
> 
> Yes, this is a new API effectively, so it needs to be added. @hakonfam please add an entry in `doc/reference/stability.rst` and also include it in the doc reference, probably in `doc/reference/storage/`

I heroically tried to add something which looked like what was already there. If something is wrong, please give me detailed instructions on how to fix it :)

hakonfam commented:
@mbolivar-nordic @pabigot Should I change the `u8_t *data` argument in `fbw_write` to be a `const u8_t *` instead of only `u8_t *`? IMO it makes sense as we do not modify the data, and users of this funtion may very well have const pointers that it wants to use.

pabigot commented:
I would prefer Zephyr to be more sensitive to using `const`-aware, so yes, if the API will never mutate the data it should be marked to indicate that fact.

hakonfam commented:
Updated the signature to `int fbw_write(struct fbw_ctx *ctx, const u8_t *data, size_t len, bool flush)`

Also updated the test to have a const write buf.

carlescufi commented:
@hakonfam apologies, the stability page has been renamed, could you rebase please?

mbolivar-nordic commented on doc/reference/overview.rst:
```diff
@@ -239,6 +239,11 @@ current :ref:`stability level <api_lifecycle>`.
 
+   * - :ref:`flash_buffered_write`
+     - Experimental
+     - 2.2
```
Shouldn't these both be 2.3?

hakonfam replied:
Probably, I had no idea what I was doing. Why is two versions required?

mbolivar-nordic replied:
Take a look here:

hakonfam replied:
And now it all makes sense, thank you :) I applied your proposal.

mbolivar-nordic reviewed (commented):
Code generally looks good. Will leave the API tags up to @pabigot since this is the trial run for his terminology. One question about the API page changes.

pabigot commented on include/storage/fbw.h:
```diff
@@ -0,0 +1,132 @@
+/**
+ * @typedef fbw_callback_t
+ *
+ * @details The fbw callback is called when data from the buffer is written
```
This is a little confusing to somebody unfamiliar with the API.  From the first sentence of the details it might appear this is invoked *to* write the data to flash, not *after*.  Something like this would be less ambiguous: state the purpose and use right away, not in the third sentence.

```

/**
 * @brief Signature for callback invoked after flash write completes.
 *
 * Functions of this type are invoked with a buffer containing
 * data read back from the flash after a flash write has completed.
 * Blah blah verifying the data...
 */
```

pabigot replied:
Also: this callback must be **isr-ok** if `fbw_flash_write()` can be invoked from an interrupt.  If you want to document API behavior.

hakonfam replied:
I have not marked `fbw_flash_write` as `isr-ok` so AFAICS the callback should not have that property either.

Added your proposal to the doc.

pabigot commented on subsys/storage/flash_buffered_write/fbw.c:
```diff
@@ -0,0 +1,192 @@
+
+
+	if (IS_ENABLED(CONFIG_FBW_ERASE)) {
+		rc = fbw_erase(ctx, write_addr + ctx->buf_bytes);
```
I think this should be `write_addr + ctx->buf_bytes - 1U`, with a check somewhere that `ctx->buf_bytes` must be positive.  Otherwise you're erasing the page containing the byte that would be written after the ctx buffer is written, which may not exist and so may return an error, preventing the ctx buffer from being written even if it would be entirely stored in the last page.

hakonfam replied:
You are correct. I added a unit test triggering this bug in addition to your fix. Since buf_bytes is type `size_t` I simply return from the function if it is `0`. The caller of this function also ensures that `buf_bytes > 0` before calling it.

pabigot commented on subsys/storage/flash_buffered_write/fbw.c:
```diff
@@ -0,0 +1,192 @@
+
+	while ((len - processed) >=
+	       (buf_empty_bytes = ctx->buf_len - ctx->buf_bytes)) {
+		memcpy(ctx->buf + ctx->buf_bytes, data + processed,
```
What's the rationale for copying the data out of the provided region in order to initiate the write, rather than just doing the write from where the data currently lives?  Is there some promise that a round-robin input buffer can wrap into the being-written data without corrupting the material written?

hakonfam replied:
There are a couple of reasons:
- Write alignment. This is not guaranteed by the drivers, and the user of `fbw` should not have to deal with it.
- Avoid synchronization issues with radio. For some HW there are restrictions on flash operations working in parallel with radio operations, the fewer flash operations the less intrusive the buffered write is to the radio performance.
- Fewer write operations will give faster average response time as seen from the app (since flash_write takes longer than memcpy).
- Fewer write operations also implies fewer calls to the driver which gives a faster application using less power
- Backwards compatibility, mcumgr transport was tested with this and there might be timing related issues for systems already using flash_img.c if they suddenly stop buffering their writes.

pabigot reviewed (commented):
I've got a concern about the way progressive erase is handled, and on the need for copying data to a local buffer to write it.  I'm also a bit skeptical of the callback-to-verify approach (why not just do memcmp of the original data provided and the data read back?).

I'm not convinced this is a good solution to the problem as I understand it, but I won't block it if other people think it meets an important need.

hakonfam commented:
>  I'm also a bit skeptical of the callback-to-verify approach (why not just do memcmp of the original data provided and the data read back?).

The `fbw` functionality will typically be used to write firmware images, which are received alongside some SHA used to verify its correctness. Having this optional callback allow users to verify the incoming firmware as it is being written to flash. Note that this can be external flash, which is why the `flash_read` operation is performed to read the data back out from where it was written.

Any `memcmp` would simply be duplicate work for the users of this module as I see it. Although one could argue that if a callback is not registered, `fbw` should perform a simple `memcmp` to give some comfort.

pabigot reviewed (dismissed):
Comments below.  I'm a bit more comfortable with what this is doing, but I'm not entirely convinced it's worth promoting to a subsystem.  At a minimum I believe the documentation needs to be extended to provide motivation and use cases so people searching for a capability can find this and understand why they might want to use it; see last part of response below.  I'm marking **request-changes** specifically to have that documentation added.  Enough to be standalone should be added in the header; more including use case descriptions can be put in flash_buffered_write.rst`.

I won't approve this because it still seems too focused on a single specific use case for a single interface (mcumgr).  I think a more flexible model would include the ability to deal with sub-image blocks rather than assuming everything comes in a contiguous sequence with no ability to back-fill lost packets.  The more I think about that gap in functionality the less happy I get, but right now I'll remove the block once the documentation makes the purpose and limitations clear.

> There are a couple of reasons:
> 
> * Write alignment. This is not guaranteed by the drivers, and the user of `fbw` should not have to deal with it.

What does "write alignment" mean?  That the buffer has some specific alignment?  There should be no requirement for that at the flash API level.  There's also no documented alignment requirement on the fbw buffer.

Does it mean that write length might need to be multiples of some minimum write size?  If so something should document that expectation and how to make sure it's satisfied, including the possibility of it not being satisfied if the image length is not a multiple of that size.

> * Avoid synchronization issues with radio. For some HW there are restrictions on flash operations working in parallel with radio operations, the fewer flash operations the less intrusive the buffered write is to the radio performance.

Ah, I see.  The use-case assumes that the amount of data received is significantly less than the allocated buffer, so multiple calls to `fbw_write()` accumulate data which is written out only when the buffer is full or the flush parameter is set.

This is not made clear in the documentation; I honestly thought the main value added was validation that the data got persisted correctly.  (Putting motivation in the PR description is not sufficient; the code should stand alone.)

> * Fewer write operations will give faster average response time as seen from the app (since flash_write takes longer than memcpy).
> * Fewer write operations also implies fewer calls to the driver which gives a faster application using less power

Agreed.

> * Backwards compatibility, mcumgr transport was tested with this and there might be timing related issues for systems already using flash_img.c if they suddenly stop buffering their writes.

This is not a compelling reason for defining a new subsystem in a way that makes it more complex and harder to use in other applications.  The motivation for adding API is that it meets the needs of multiple users, not just the one from which the implementation derives.

> > I'm also a bit skeptical of the callback-to-verify approach (why not just do memcmp of the original data provided and the data read back?).

I was assuming the incoming block was larger than the buffer being used to write and verify the write, in which case this API makes no sense.

Somewhere in all this there should be a clear description of what this subsystem is intended to do.  I now understand that to be: aggregate incoming sequential blocks of data into a contiguous block that can be written to a flash device efficiently.

There should also be a description of when you'd want to use it: when receiving a data stream over a transport that carries small payloads.

And what it gives you: reduced number of writes, an opportunity to avoid differences in transaction size requirements between the transport and the flash device, the option of doing stream validation of the data in flash-size blocks, ....

hakonfam commented:
> I think a more flexible model would include the ability to deal with sub-image blocks rather than assuming everything comes in a contiguous sequence with no ability to back-fill lost packets.

I agree, the problem is that I don't have a use case for this at hand, and I don't want to go ahead and design an API without a use case driving the design. My plan was to add this once I had some use case which needed it, and try to design the current API to facilitate this extension in the future.

It might be that the unit test could be the use case, but I doubted that this would get accepted.

hakonfam commented on include/storage/fbw.h:
```diff
@@ -0,0 +1,133 @@
+ *
+ * @return non-negative on success, negative errno code on fail
+ */
+int fbw_erase(struct fbw_ctx *ctx, off_t off);
```
I agree

hakonfam commented on subsys/storage/flash_buffered_write/fbw.c:
```diff
@@ -0,0 +1,196 @@
+	}
+
+	if (ctx->callback) {
+		rc = flash_read(ctx->fdev, write_addr, ctx->buf,
```
I added the following 

```
		/* Invert to ensure that caller is able to discover a faulty
		 * flash_read() even if no error code is returned.
		 */
		for (int i = 0; i < ctx->buf_bytes; i++) {
			ctx->buf[i] = ~ctx->buf[i];
		}

```

Do you think it is sufficient to invert the number of bytes that were being written, or should we invert the entire buffer no matter what?

pabigot replied:
The region being read back is sufficient.

hakonfam commented:
@pabigot I added the docs as you proposed offline (let me know if you want more credit) in addition to some motivation for when you might want to use this.

pabigot reviewed (commented):
There's enough documentation of the motivation for and use of this that I've removed my NAK; thank you.

I still don't feel this has sufficient value to be a top-level API in Zephyr, so I won't approve it, but it has two approvals already so that doesn't matter.  Below are my out-of-band comments explaining what I would have liked.  I agree that doing this is a lot of work for no clear benefit, but that only supports my position that this doesn't need to be an API.

> OK, so let me step back and do this for real.
>
> Let's start with what might be the description of the API:
> 
> fsw takes contiguous fragments of a stream of data (e.g. from radio packets), aggregates them into a user-provided buffer, then when the buffer fills (or stream ends) writes it to a raw flash partition.  It supports providing the read-back buffer to the client to use in validating the persisted stream content.
>  
> Now let's ask questions.  There are two obvious ones.
>  
> First, this is only useful if the stream is being persisted to raw flash.  What if I'm receiving an image I want to store in a littlefs file system, and I'll write it to the mcuboot partition when I want to switch the target to that particular role?  I still need to aggregate stream fragments for optimized write and validation, but I can't use this API because I need to write using the fs API not the flash API.  Maybe the API should just provide a callback with the aggregated data, and let the client handle how the larger blocks are processed.
>
> Second, there are two basic functions: add a fragment to the buffer, and process a completed buffer.  The as-implemented solution invokes the second from the first, which requires that both are performed by a single thread, with the possibility that the processing phase will cause the thread to sleep.  This means I can't append fragments from an ISR, and I have make sure whatever appends fragments isn't a critical thread that has to be responsive to other events.
>
> Dealing with the second is probably a lot of work: you'd need locks, a hand-off to a thread to processes the buffer, and figure out what to do if another fragment comes in while the buffer is full.  So it's reasonable to say that's out of scope.
>
> But the first seems much more useful and simple to accommodate: decouple the aggregation from the flash, perhaps providing two APIs:
>  
> * The stream collector data structure accepts fragments of data and aggregates them into a user-provided buffer, and synchronously invokes a user-provided callback providing the full (or partial last) buffer. (There may already be a data structure like this in Zephyr.)
>  
> * The flash stream writer pairs a stream collector with flash configuration and pre-defined functions that write completed buffers to flash and provides the read-back data to the client for external validation.
>  
> So: Yes, renaming it to flash stream writer is good, but to solve the general case it needs to be decoupled from the destination storage solution.  I would be much more supportive of a general API than one that only satisfies a single application need.

hakonfam commented:
Sorry for the spam, but ping @carlescufi @mbolivar-nordic @nvlsianpu 

I need an answer to if this is going in or not. If its not going in I need to start the process of getting it in our downstream.

nvlsianpu commented:
I already approved, @carlescufi need your voice here

carlescufi commented:
@pabigot and @mbolivar-nordic I'd really like to get this into 2.3. Is there a chance we can move this forward?

carlescufi commented:
@de-nordic could you review please?

hakonfam commented:
@pabigot I will post a revised version of this with your proposed changes at a later time. I made a draft of this to ensure that the API exposed now will not change when the generalized version is in place.

de-nordic reviewed (approved):
I think it is fine.

de-nordic committed:
west.yml: Snapshot update of mcumgr from upstream

This commit updates mcumgr with latest snapshot from upstream.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>
